sys/include/net/gcoap.h

```diff
@@ -378,12 +381,14 @@ typedef struct {
  * @brief  Container for the state of gcoap itself
  */
 typedef struct {
+    mutex_t lock;                       /**< Shares state attributes safely */
     gcoap_listener_t *listeners;        /**< List of registered listeners */
     gcoap_request_memo_t open_reqs[GCOAP_REQ_WAITING_MAX];
                                         /**< Storage for open requests; if first
                                              byte of an entry is zero, the entry
                                              is available */
-    uint16_t last_message_id;           /**< Last message ID used */
+    atomic_uint_least16_t next_message_id;
```

Comment:
Ah, I see. I am a bit cautious about using the int_leastxx types because of the risk of forgetting that for example sizeof might give a different answer than you expect for a 16 bit value, and that you no longer get the implicit bitwise AND 0xffff on all values. There may be hidden bugs which will only show up on the platforms which have sizeof uint_least16_t > 2.